ipn/ipnlocal/local.go

```diff
@@ -4837,9 +4840,45 @@ func (b *LocalBackend) doSetHostinfoFilterServices() {
 	c := len(hi.Services)
 	hi.Services = append(hi.Services[:c:c], peerAPIServices...)
 	hi.PushDeviceToken = b.pushDeviceToken.Load()
+
+	if hostinfoPeerAPIPortsChanged(oldServices, hi.Services) {
+		b.logf("Hostinfo peerAPI ports changed: was %v, now %v", oldServices, hi.Services)
+	}
+
 	cc.SetHostinfo(&hi)
 }
 
+// hostinfoPeerAPIPortsChanged compares the peerAPI ports between two slices of services.
+// It returns true if the ports for either "peerapi4" or "peerapi6" have changed.
+func hostinfoPeerAPIPortsChanged(oldServices, newServices []tailcfg.Service) bool {
+	oldMap := make(map[string]uint16)
+	newMap := make(map[string]uint16)
+
+	for _, s := range oldServices {
+		if s.Proto == "peerapi4" || s.Proto == "peerapi6" {
+			oldMap[string(s.Proto)] = s.Port
+		}
+	}
+	for _, s := range newServices {
+		if s.Proto == "peerapi4" || s.Proto == "peerapi6" {
+			newMap[string(s.Proto)] = s.Port
+		}
+	}
+
+	if len(oldMap) != len(newMap) {
+		return true
+	}
+
+	// Compare the port numbers for each peerAPI protocol.
+	for proto, oldPort := range oldMap {
+		if newPort, ok := newMap[proto]; !ok || newPort != oldPort {
+			return true
+		}
+	}
```

Comment:
maps.Equal ?

But we could do this without allocating maps at all. There are only 4 uint16s to track:

```go
    type pair struct {   v4, v6 uint16   }
    var portsOld, portsNew pair
    pop := func(p *pair, ss []tailcfg.Service) {
       for _, s := range ss {
             switch s.Proto {
                  case "peerapi4": p.v4 = s.Port
                  case "peerapi6": p.v6 = s.Port
            }
       }
   }
   pop(&portsOld, oldServices)
   pop(&portsNew, newServices)
   return portsOld == portsNew
```

Reply:
And then for the logging big, you could return log or return the port values from that?